R/rowWilcoxonTests.R

```diff
@@ -72,26 +71,152 @@
 #' cls_perm <- replicate(B, sample(cls))
 #' system.time(fwt <- rowWilcoxonTests(mat, categ = cls_perm, alternative = "two.sided"))
 #' 
-rowWilcoxonTests <- function(mat, categ, 
-                              alternative = c("two.sided", "less", "greater"), 
-                              correct = TRUE) {
+
+rowWilcoxonTests <- function (mat, categ, alternative = c("two.sided", "less", "greater"), 
+                                correct = TRUE) 
+{
+    alternative <- match.arg(alternative)
+    stopifnot(all(categ %in% c(0, 1)))
+    categ <- as.matrix(categ)
+    levels(categ) <- NULL
+    
+    rks <- rowRanks(mat, ties.method = "average")
+    ny <- sum(categ[,1] == 0) 
+    nx <- sum(categ[,1] == 1) 
+    min_stat <- nx * (nx + 1)/2 
+    
+    stats <- rks %*% categ
+    stats <- stats - min_stat
+    
+    rks2 <- rks 
+    mode(rks2) <- "integer"
+    n_ties <- rowTabulates(rks2) 
+    ties <- rowSums(n_ties^3 - n_ties) 
+    sigma <- sqrt((nx * ny/12) * ((ny + nx + 1) - ties/((ny + nx) * (ny + nx - 1)))) 
+    z <- stats - ny * nx/2
+    CORRECTION <- 0
+    if (correct) {
+        CORRECTION <- switch(alternative, two.sided = sign(z) * 
+                                 0.5, greater = 0.5, less = -0.5)
+    }
+    
+    z <- (z - CORRECTION)/sigma
+    p <- switch(alternative, 
+                less = pnorm(z), 
+                greater = pnorm(z, lower.tail = FALSE), 
+                two.sided = 2 * pmin(pnorm(z), pnorm(z, lower.tail = FALSE)))
+    
+    B <- ncol(categ)
+    m <- nrow(mat)
+    est <- matrix(NA_real_, nrow = m, ncol = B)
+    if (dim(categ)[2] == 1){
+        wx <- which(categ == 1)
+        est <- rowMedians(mat[, wx]) - rowMedians(mat[, -wx])
+        stats <- as.vector(stats)
+        p <- as.vector(p)
+    } 
+    list(p.value = p, statistic = stats, estimate = est)
+}
+
+
+# Version 2 of Wilcoxon test. 
+# In this version, all calculations that can be done outside the loop for permutations are processed upstream.
+
+
+
+rowWilcoxonTestsV2 <- function (mat, categ, alternative = c("two.sided", "less", "greater"), 
+                                correct = TRUE) 
+{
+    alternative <- match.arg(alternative)
+    stopifnot(all(categ %in% c(0, 1)))
+    categ <- as.matrix(categ)
+    levels(categ) <- NULL
+    
+    
+    rks <- rowRanks(mat, ties.method = "average")
+    ny <- sum(categ[,1] == 0) 
+    nx <- sum(categ[,1] == 1) 
```

Comment:
Same comment as for `rowWilcoxonTests` (the check may be here or in the loop on `b` below).

Reply:
This comment should be ignored if the entire function is dropped!